src/GafferSceneUI/ScriptNodeAlgo.cpp

```diff
@@ -0,0 +1,150 @@
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferSceneUI/ScriptNodeAlgo.h"
+
+#include "GafferSceneUI/ContextAlgo.h"
+
+#include "Gaffer/Context.h"
+#include "Gaffer/ScriptNode.h"
+
+#include "boost/bind/bind.hpp"
+
+#include <unordered_map>
+
+using namespace boost::placeholders;
+using namespace Gaffer;
+using namespace GafferSceneUI;
+
+namespace
+{
+
+struct ChangedSignals
+{
+	ScriptNodeAlgo::ChangedSignal visibleSetChangedSignal;
+	ScriptNodeAlgo::ChangedSignal selectedPathsChangedSignal;
+	Gaffer::Signals::ScopedConnection connection;
+};
+
+void contextChanged( IECore::InternedString variable, ScriptNode *script, ChangedSignals *signals )
+{
+	if( ContextAlgo::affectsVisibleSet( variable ) )
+	{
+		signals->visibleSetChangedSignal( script );
+	}
+
+	if( ContextAlgo::affectsSelectedPaths( variable ) || ContextAlgo::affectsLastSelectedPath( variable ) )
+	{
+		signals->selectedPathsChangedSignal( script );
+	}
+}
+
+ChangedSignals &changedSignals( ScriptNode *script )
+{
+	static std::unordered_map<const ScriptNode *, ChangedSignals> g_signals;
+	ChangedSignals &result = g_signals[script];
+	if( !result.connection.connected() )
+	{
+		// Either we just made the signals, or an old ScriptNode
+		// was destroyed and a new one made in its place.
+		result.connection = const_cast<Context *>( script->context() )->changedSignal().connect(
+			boost::bind( &contextChanged, ::_2, script, &result )
+		);
+	}
+	return result;
+}
+
+} // namespace
+
+/// Everything here is implemented as a shim on top of ContextAlgo. Our intention is to move everyone
+/// over to using ScriptNodeAlgo, then to remove ContextAlgo and reimplement ScriptNodeAlgo using the
+/// metadata API to store the state as metadata on the ScriptNOde. This will bring several benefits :
```

Comment:
ScriptNOde -> ScriptNode

Reply:
Thanks for spotting that - I've squashed the fix in and am merging...